### PR TITLE
[mypy] [9880] Reduce mypy errors

### PR DESCRIPTION
**Pull request by rodrigc**
https://twistedmatrix.com/trac/ticket/9880

### CODE REVIEW

File: src/twisted/internet/interfaces.py [hunk 1]
```diff
@@ -1243,14 +1250,14 @@ def seconds():
         """
 
 
-    def callLater(delay, callable, *args, **kw):
+    def callLater(delay, f: Callable[..., Any], *args, **kw):
```
Comment (adiroiban) [inline src/twisted/internet/interfaces.py]: I don't know what to say about this interface change.

@glyph  do you have any suggestion?

I would way that this change should be made in a separate PR.

I don't know if `f` is the best name ... maybe use `func`

```
$ grep -r "def callLater(" src/
src/twisted/protocols/policies.py:    def callLater(self, period, func):
src/twisted/protocols/policies.py:    def callLater(self, period, func):
src/twisted/protocols/policies.py:    def callLater(self, period, func):
src/twisted/test/test_policies.py:    def callLater(self, period, func):
src/twisted/test/test_policies.py:    def callLater(self, period, func):
src/twisted/test/test_policies.py:    def callLater(self, timeout, func, *args, **kwargs):
src/twisted/names/dns.py:    def callLater(self, period, func, *args):
src/twisted/conch/test/test_helper.py:    def callLater(self, timeout, f, *a, **kw):
src/twisted/web/http.py:    def callLater(self):
src/twisted/web/http.py:    def callLater(self, value):
src/twisted/internet/_threadedselect.py:    def callLater(self, *args, **kw):
src/twisted/internet/interfaces.py:    def callLater(delay, callable, *args, **kw):
src/twisted/internet/asyncioreactor.py:    def callLater(self, seconds, f, *args, **kwargs):
src/twisted/internet/cfreactor.py:    def callLater(self, _seconds, _f, *args, **kw):
src/twisted/internet/task.py:    def callLater(self, when, what, *a, **kw):
src/twisted/internet/base.py:    def callLater(self, _seconds, _f, *args, **kw):
src/twisted/internet/_glibbase.py:    def callLater(self, *args, **kwargs):
```

but is best to keep the scope of this PR to only type hinting and make the backward incompatible changes in a separate PR.

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thanks. This looks good.

My request it to remove the backward incompatible changes from this PR and have them in a separate PR dedicated to changing the signature of the interfaces.

**Comment by glyph** [comment]
> My request it to remove the backward incompatible changes from this PR and have them in a separate PR dedicated to changing the signature of the interfaces.

FWIW: very much agreed on this point :)

**Comment by rodrigc** [comment]
> > My request it to remove the backward incompatible changes from this PR and have them in a separate PR dedicated to changing the signature of the interfaces.
> 
> FWIW: very much agreed on this point :)


@glyph @adiroiban OK I submitted a separate PR for this: https://github.com/twisted/twisted/pull/1341

File: src/twisted/internet/interfaces.py [hunk 2]
```diff
@@ -1340,7 +1347,7 @@ class IReactorFromThreads(Interface):
     @since: 15.4
     """
 
-    def callFromThread(callable, *args, **kw):
+    def callFromThread(f: Callable[..., Any], *args, **kw):
```
Comment (adiroiban) [inline src/twisted/internet/interfaces.py]: This is still an interface change callable -> f

File: src/twisted/internet/interfaces.py [hunk 3]
```diff
@@ -1364,7 +1371,7 @@ class IReactorInThreads(Interface):
     @since: 15.4
     """
 
-    def callInThread(callable, *args, **kwargs):
+    def callInThread(f: Callable[..., Any], *args, **kwargs):
```
Comment (adiroiban) [inline src/twisted/internet/interfaces.py]: This is still an interface change callable -> f

File: src/twisted/internet/interfaces.py [hunk 4]
```diff
@@ -1508,15 +1517,15 @@ def removeSystemEventTrigger(triggerID):
             C{triggerID}.
         """
 
-    def callWhenRunning(callable, *args, **kw):
+    def callWhenRunning(f: Callable[..., Any], *args, **kw):
```
Comment (adiroiban) [inline src/twisted/internet/interfaces.py]: This is still an interface change callable -> f

File: src/twisted/internet/task.py [hunk 5]
```diff
@@ -786,15 +787,15 @@ def _sortCalls(self):
         self.calls.sort(key=lambda a: a.getTime())
 
 
-    def callLater(self, when, what, *a, **kw):
+    def callLater(self, delay, f: Callable[..., Any], *args, **kw):
```
Comment (adiroiban) [inline src/twisted/internet/task.py]: should keep `when` and `what` as they are.

The code from below is valid code, but will break after this change

```python
reactor.callLater(when=0.3, what=recover_response)
```

File: src/twisted/internet/interfaces.py (hunk 1 above)
Comment (adiroiban) [reply src/twisted/internet/interfaces.py]: This is still an interface change callable -> f

File: src/twisted/internet/base.py [hunk 6]
```diff
@@ -719,14 +719,15 @@ def fireSystemEvent(self, eventType):
             event.fireEvent()
 
 
-    def addSystemEventTrigger(self, _phase, _eventType, _f, *args, **kw):
+    def addSystemEventTrigger(self, phase: str, eventType: str,
+                              f: Callable[..., Any], *args, **kw):
```
Comment (adiroiban) [inline src/twisted/internet/base.py]: This is still a signature change _f -> f 

But since _f is a "private" we can change it to public... but if it should be change, it should be change to `callable` to match the interface definition.

The interface is defined as
```
src/twisted/internet/interfaces.py:    def addSystemEventTrigger(phase, eventType, callable, *args, **kw):
```

File: src/twisted/internet/base.py [hunk 7]
```diff
@@ -736,14 +737,14 @@ def removeSystemEventTrigger(self, triggerID):
         self._eventTriggers[eventType].removeTrigger(handle)
 
 
-    def callWhenRunning(self, _callable, *args, **kw):
+    def callWhenRunning(self, f: Callable[..., Any], *args, **kw):
```
Comment (adiroiban) [inline src/twisted/internet/base.py]: this is a signature change same as addSystemEventTrigger ... this should be named `callable`

File: src/twisted/internet/test/test_base.py [hunk 8]
```diff
@@ -71,7 +72,7 @@ def seconds(self):
         pass
 
 
-    def callInThread(self, callable, *args, **kwargs):
+    def callInThread(self, f: Callable[..., Any], *args, **kwargs):
```
Comment (adiroiban) [inline src/twisted/internet/test/test_base.py]: this is a signature change.

File: src/twisted/internet/testing.py [hunk 9]
```diff
@@ -577,15 +578,16 @@ def fireSystemEvent(self, eventType):
         raise NotImplementedError()
 
 
-    def addSystemEventTrigger(self, phase, eventType, callable, *args, **kw):
+    def addSystemEventTrigger(self, phase: str, eventType: str,
+                              f: Callable[..., Any], *args, **kw):
```
Comment (adiroiban) [inline src/twisted/internet/testing.py]: I guess that for the testing code this is ok, but is better to have this in the separate PR.

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thanks for the update.
I still see changes to the interface and method signature.

File: src/twisted/internet/test/test_base.py (hunk 8 above)
Comment (rodrigc) [reply src/twisted/internet/test/test_base.py]: I took this signature change out

File: src/twisted/internet/base.py (hunk 7 above)
Comment (rodrigc) [reply src/twisted/internet/base.py]: I took this signature change out

File: src/twisted/internet/base.py (hunk 6 above)
Comment (rodrigc) [reply src/twisted/internet/base.py]: I took this signature change out

File: src/twisted/internet/task.py (hunk 5 above)
Comment (rodrigc) [reply src/twisted/internet/task.py]: I took this out

File: src/twisted/internet/interfaces.py (hunk 4 above)
Comment (rodrigc) [reply src/twisted/internet/interfaces.py]: I took this out

File: src/twisted/internet/interfaces.py (hunk 3 above)
Comment (rodrigc) [reply src/twisted/internet/interfaces.py]: I took this out

File: src/twisted/internet/interfaces.py (hunk 2 above)
Comment (rodrigc) [reply src/twisted/internet/interfaces.py]: I took this out

File: src/twisted/internet/testing.py (hunk 9 above)
Comment (rodrigc) [reply src/twisted/internet/testing.py]: I took this out

**Comment by rodrigc** [comment]
@adiroiban I changed this PR to just focus on removing default arguments from interfaces, and adding type annotations to some method arguments

**Review by adiroiban — APPROVED** [review]
Wow. Great work. Thanks.

I see that mypy is not complaining about the callable argument.

I have create the following tests and mypy is ok with the file

```python
from zope.interface import Interface, Attribute, implementer

class ITest(Interface):
    test1 = Attribute('test1')
    def test2(callable, two):
        "Order or arguments."

@implementer(ITest)
class Test:
    def __init__(self):
        self.test1 = 1

    def test2(self, callable, one):
        """
        Mixed order.
        """
```
